README.md
# Tyrant's Virtual Mission Tools (VMT)
A mission planning tool for Falcon BMS. Visualize the battlefield, generate weather, and more.

Successor to [WeatherGen](https://candera.github.io/weathergen).

# Installation and Instructions
Visit the [BMS forum thread](https://www.bmsforum.org/forum/showthread.php?31611-Release-Tyrant-s-Virtual-Mission-Tools-(VMT)) for instructions and download link.

# License

MIT License

```
Copyright (c) 2017 Craig Andera

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```



changelog.org
#+OPTIONS: toc:nil
#+OPTIONS: num:nil
#+OPTIONS: html-postamble:nil
#+TITLE:

* Change Log
** v063-5 [2021-07-17 Sat]
- Bugfix: Kuwait missions wouldn't load due to gaps in XML database indicies
- Bugfix: Null images were preventing mission load
- Give better error for non-1024x1024 theaters

** v063 [2021-04-06 Tue]
- Bugfix: PPT symbols would not always display
- Remove 32-bit versions, as I can't get them to build any more
** v062 [2021-03-07 Sun]

- Bugfix: Saving individual files (map images, annotations, FMAPs,
  etc.) would silently fail

** v061 [2021-03-07 Sun]

- Bugfix: correct 'no protocol method ISpec.read defined for type
  function' error when loading a TE

** v060 [2021-01-26 Tue]

Update to support BMS 4.35.

THIS VERSION WILL WORK ONLY WITH BMS RELEASE 4.35.

Looking for a 4.34-compatible version of VMT? Download the [[https://s3.amazonaws.com/org.craigandera.vmt/dev/VMT-win32-x64-v058-19-gf8e70f4-dev.zip][32-bit]] or
[[https://s3.amazonaws.com/org.craigandera.vmt/dev/VMT-win32-ia64-v058-19-gf8e70f4-dev.zip][64-bit]] versions.

Changes:

- Added ability to set mission date in weather settings.

Note that because of the complexity in replicating BMS turbulence
calculations, turbulence will always be set to the default when saving
a TWX file. You will need to edit it in Falcon after saving if you
want non-default turbulence behavior.

** v058-22 [2019-08-20 Tue]

DIAGNOSTIC RELEASE.

- Adds "Open Debug Console" button to Tools tab
- Shows mission window even if loading fails

These features may be removed in a future release - don't count on
them sticking around. They are there to support debugging an issue
that only shows up on some users' Windows 10 installs.

** v058-19 [2019-07-03 Wed]

BETA Release!

- Fix the stupid bugs I had in the last beta: FMAP and TWX saving
  should work now.
- Fix app icons.
- Allow overriding visibility and temperature.

** v058-16 [2019-07-01 Mon]

BETA Release!

- Weather should be working now.
- Implement new and improved weather override regions: override wind,
  weather type, temperature, and/or clouds.
- Remove wind stability regions, as these are subsumed by improved
  overrides.
- Known bug: occasionally weather parameters do not update the model.

** v058-10 [2019-06-08 Sat]

ALPHA Release! Weather generation does not work yet! Everything else
should, but please report bugs.

- Fixed bug where active teams were not computed correctly.

** v058-9 [2019-06-02 Sun]

ALPHA Release! Weather generation does not work yet! Everything else
should, but please report bugs.

- Fixed bug where weapons list in the damage computer was missing
  weapons.
- Remove warning about squadrons with no airbase - this is apparently
  no longer a problem in 4.34.

** v058-6 [2019-06-02 Sun]
- ALPHA Release: Update to support 4.34
  - Weather non-functional
  - Other bugs likely present
  - 4.33 no longer supported: use v058 or prior
** v058-2 [2018-09-20 Thu]
- Remove highlight from flight when closing flight tab.
- Make tools tab always visible.
** v058 [2018-09-18 Tue]
- Fix typos in download page.
- Build Linux version as well
- Change logo to 440th VFW's
- Fix various formatting problems
- Fix incorrect label on annotations copy button
- Improved map sizing and positioning: controls will never move below
  map. Instead, map gets smaller if necessary to accommodate controls
  when resizing window.
- Titlebar of map is collapsible to increase usable space.
- Tab names now remain visible even when scrolling content.
- Fixed width of dropdowns in annotation controls.
- Implemented "Save map image as PNG"
- Fixed intermittent bug where things wouldn't render right sometimes
- Flights can be opened in their own tabs
- Fix (again) high CPU when app is idle
- Flight colors and check state are now saved with briefing
** v057 [2018-02-01 Thu]
- Add missing help text
** v056 [2018-01-28 Sun]
- Annotations!
  - Add arrows, text, shapes, and lines to a map and save them with a
    briefing
- Can now save modified briefing when viewing a briefing
- Fix bug where dragging off the map makes things jump around
- Reverse use of triangles/circles in waypoints to match BMS
- Include download link in update notifications.
** v055 [2017-12-26 Tue]
- Fix bug where sorting was broken on flights page.
- Fix bug where flight color selection was not working.
- Fix occasional runaway CPU utilization.
** v054 [2017-12-22 Fri]
- Fix minor bug where "VMT Settings" text was showing when it shouldn't.
** v053 [2017-12-21 Thu]
- Add ability to have VMT automatically check for updated versions of
  itself.
- Fix problem where web links were opening in the app instead of in a
  browser.
** v052 [2017-12-18 Mon]
- Major speedup in mission and briefing loading.
- Add checkbox in map controls section allowing show/hide of borders
  of wind stability areas.
- Wind stability areas and weather override regions are automatically
  placed in edit mode when created.
- Fix bug introduced in v051 where squadrons weren't showing up for
  carriers and army bases.
- Fix bug where weather override edit button wasn't displaying
  correctly when editing.

** v051 [2017-12-16 Sat]
- Handle [[https://www.bmsforum.org/forum/showthread.php?31611-Release-Tyrant-s-Virtual-Mission-Tools-(VMT)&p=449550&viewfull=1#post449550][weird Balkans bad index bug]].
- Report warning when squadrons have no airbase
- Handle problem with improper handling of time in briefing files.
** v050 [2017-11-18 Sat]
- Fix bug handling consecutive steerpoints with the same time
- Fix bug where briefing wouldn't load due to nil time somewhere.
** v049 [2017-10-15 Sun]
- Fix bug where install ID was saved incorrectly in briefing files.
- Save version of VMT used to create briefing files; warn on load if a
  different version of VMT is being used.
** v048 [2017-09-30 Sat]
- Briefing notes: a simple (for now) text box where mission designers
  can put notes to be displayed when the briefing file is opened.

dev/user.clj
(ns user
  (:require [clojure.java.io :as io]
            [clojure.pprint :as pprint :refer [pprint]]
            [clojure.repl :refer :all]
            [clojure.set :as set]
            [clojure.spec.alpha :as s]
            [clojure.spec.test.alpha :as stest]
            [clojure.string :as str]
            [clojure.tools.namespace.repl :refer [refresh]]
            [cognitect.transit :as transit]
            [expound.alpha :as expound]
            [garden.core :refer [css]]
            [garden.selectors :as css-sel]
            [lsobot.acmi :as acmi]
            [octet.core :as buf]
            [taoensso.timbre :as log]
            [weathergen.compression :as compress]
            [weathergen.coordinates :as coords]
            [weathergen.dtc :as dtc]
            [weathergen.encoding :as encoding]
            [weathergen.fmap :as fmap]
            [weathergen.falcon.constants :as c]
            [weathergen.falcon.files.mission :as mission]
            [weathergen.falcon.files.mission.xml :as xml]
            [weathergen.falcon.files.resources :as resources]
            [weathergen.filesystem :as fs]
            [weathergen.falcon.files.images :as im]
            [weathergen.lzss :as lzss]
            [weathergen.math :as math]
            [weathergen.model :as model]
            [weathergen.time :as time]
            [weathergen.util :as util]))

#_(defn cljs
  []
  (start-figwheel!)
  (cljs-repl))

#_(in-ns 'boot.user)

(def installs (-> "/Users/candera/Library/Application Support/craigandera.org/VMT/settings.edn"
                  slurp
                  read-string
                  :installations))

(def smpu-old
  (delay
   (log/debug "Reading SMPU...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Campaign/SAVE/SMPU-Day  3 00 41 48.cam")]
     (log/debug "Done reading.")
     mission)))

(def smpu
  (delay
   (log/debug "Reading SMPU...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Add-On Korea EM1989 v2/Campaign/SMPU-Day  1 23 55 48.cam")]
     (log/debug "Done reading.")
     mission)))

(def save2
  (delay
   (log/debug "Reading save2...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Campaign/SAVE/save2.cam")]
     (log/debug "Done reading.")
     mission)))

(def stratus-te
  (delay
   (log/debug "Reading stratus.tac...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Campaign/SAVE/stratus.tac")]
     (log/debug "Done reading.")
     mission)))

(def te-new
  (delay
   (log/debug "Reading te_new.tac...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Campaign/SAVE/te_new.tac")]
     (log/debug "Done reading.")
     mission)))

(def wnpu
  (delay
   (log/debug "Reading WNPU...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Add-On Korea Strong DPRK/Campaign/WNPU-Day  1 21 07 12.cam")]
     (log/debug "Done reading.")
     mission)))

(def ito
  (delay
   (log/debug "Reading ITO...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Add-On Israel/campaign/Solid Truss Start.cam")]
     (log/debug "Done reading.")
     mission)))

(def balkans
  (delay
   (log/debug "Reading Balkans...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Add-On Balkans/Campaign/Powder Keg Start.cam")]
     (log/debug "Done reading.")
     mission)))

(def fnpu
  (delay
   (log/debug "Reading FNPU...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Campaign/SAVE/FNPU-Day  1 05 12 15.cam")]
     (log/debug "Done reading.")
     mission)))

(def ffw
  (delay
   (log/debug "Reading Custom 1st VFW theater...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.33.3/Data/Add-On 1st VFW/Campaign/Save/Rolling Fire Start.cam")]
     (log/debug "Done reading.")
     mission)))

(def mantis
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/MantisAtDawn-Day  1 18 30 05.cam")]
     (log/debug "Done reading.")
     mission)))

(def tiger-spirit
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/Tiger Spirit-Day 1 08 15 42.cam")]
     (log/debug "Done reading.")
     mission)))

(def iron-fortress
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/Iron Fortress-Day  1 08 00 04.cam")]
     (log/debug "Done reading.")
     mission)))

(def rolling-fire
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/Rolling Fire-Day  1 05 00 07.cam")]
     (log/debug "Done reading.")
     mission)))


(def bear-trap
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/Bear Trap-Day  1 12 00 04.cam")]
     (log/debug "Done reading.")
     mission)))

(def double-dragon
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/Double Dragon-Day  1 16 00 04.cam")]
     (log/debug "Done reading.")
     mission)))

(def generic-te
  (delay
   (log/debug "Reading 4.34 mission...")
   (let [mission (mission/read-mission
                  installs
                  "/Users/candera/falcon/4.34/Data/Campaign/generic.tac")]
     (log/debug "Done reading.")
     mission)))

(defn smoke-test
  []
  (map :mission-name [@balkans @ito @wnpu @te-new @stratus-te @save2 @smpu @smpu-old]))

(defn read-briefing
  [path]
  (-> path
      io/input-stream
      java.util.zip.GZIPInputStream.
      (transit/reader :json)
      transit/read))

(defn dump-mission
  [mission dir]
  (->> mission
       :class-table
       util/csv-ize
       (spit (fs/path-combine dir "class-table.csv")))
  (->> mission
       :vehicle-class-data
       util/csv-ize
       (spit (fs/path-combine dir "vehicle-classes.csv")))
  (->> mission
       :objectives
       util/csv-ize
       (spit (fs/path-combine dir "objectives.csv")))
  (->> mission
       :objective-class-data
       util/csv-ize
       (spit (fs/path-combine dir "objective-classes.csv")))
  (->> mission
       :units
       util/csv-ize
       (spit (fs/path-combine dir "units.csv")))
  (->> mission
       :unit-class-data
       util/csv-ize
       (spit (fs/path-combine dir "unit-classes.csv"))))

(defn expound-explainer
  [data]
  (if (nil? data)
    (println "Success!")
    (expound/printer data)))

(defn rebl
  []
  (require '[cognitect.rebl])
  (let [f (resolve 'cognitect.rebl/ui)]
    (f)))

(defn inspect
  [x]
  (require '[cognitect.rebl])
  (eval `(cognitect.rebl/inspect ~x)))

(defmacro inspect
  [x]
  `(do
     (require '[cognitect.rebl])
     (cognitect.rebl/inspect ~x)))

#_(set! s/*explain-out* expound/printer)
#_(alter-var-root #'s/*explain-out* expound-explainer)

#_(stest/instrument)

(def weather-params
    {:temp-uniformity      0.7
   :pressure             {:min 28 :max 31 :variance 1.2 :speed 100}
   ;; From Ahmed, cell count is a function of theater size:
   ;; #define CELLSIZE  57344
   ;; #define FMAP_CELLS_FROM_SIZE(s) round(floor((s*1000)*3.28084 / (double)CELLSIZE +1))
   :cell-count           [59 59]
   :feature-size         10
   :categories           {:sunny     {:weight     50
                                      :wind       {:min 5 :mean 10 :max 30}
                                      :temp       {:min 20 :mean 22 :max 24}
                                      :visibility {:from 15 :to 30}}
                          :fair      {:weight     50
                                      :wind       {:min 0 :mean 7 :max 20}
                                      :temp       {:min 18 :mean 21 :max 23}
                                      :visibility {:from 10 :to 30}
                                      :low-clouds {:base     {:from 3000 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :few :to :broken}
                                                   :towering 0.48}}
                          :poor      {:weight     50
                                      :wind       {:min 10 :mean 15 :max 30}
                                      :temp       {:min 15 :mean 18 :max 21}
                                      :visibility {:from 2 :to 10}
                                      :low-clouds {:base     {:from 0 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :scattered
                                                              :to   :overcast}
                                                   :towering 0.35}}
                          :inclement {:weight     50
                                      :wind       {:min 15 :mean 25 :max 60}
                                      :temp       {:min 12 :mean 14 :max 16}
                                      :visibility {:from 15 :to 30}
                                      :low-clouds {:base     {:from 0 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :scattered
                                                              :to   :overcast}
                                                   :towering 0.25}}}
   :winds-aloft          {3000  {:speed {:from 2 :to 3}
                                 :bias  0.1}
                          6000  {:speed {:from 4 :to 6}
                                 :bias  0.2}
                          9000  {:speed {:from 7 :to 9}
                                 :bias  0.3}
                          12000 {:speed {:from 8 :to 12}
                                 :bias  0.4}
                          18000 {:speed {:from 11 :to 13}
                                 :bias  0.5}
                          24000 {:speed {:from 13 :to 17}
                                 :bias  0.6}
                          30000 {:speed {:from 16 :to 18}
                                 :bias  0.7}
                          40000 {:speed {:from 18 :to 22}
                                 :bias  0.8}
                          50000 {:speed {:from 20 :to 25}
                                 :bias  0.9}}
   :turbulence           {:size 1 :power 250}
   :origin               [1000 1000]
   :evolution            3600
   :time                 {:offset  1234
                          :current {:day 1 :hour 5 :minute 0}
                          :max     nil}
   :wind-uniformity      0.7
   :crossfade            0.1
   :prevailing-wind      {:heading 325}
   :seed                 1234
   :weather-overrides    [#_{:location               {:x 22
                                                      :y 45}
                             :radius                 10
                             :falloff                5
                             :animate?               false
                             :begin                  {:day 1 :hour 5 :minute 0}
                             :peak                   {:day 1 :hour 6 :minute 0}
                             :taper                  {:day 1 :hour 8 :minute 0}
                             :end                    {:day 1 :hour 9 :minute 0}
                             :pressure               28.5
                             :strength               1
                             :show-outline?          false
                             :exclude-from-forecast? false
                             :editing?               false}]}
  #_{:temp-uniformity      0.7
   :pressure             {:min 28 :max 31 :variance 1.2 :speed 100}
   ;; From Ahmed, cell count is a function of theater size:
   ;; #define CELLSIZE  57344
   ;; #define FMAP_CELLS_FROM_SIZE(s) round(floor((s*1000)*3.28084 / (double)CELLSIZE +1))
   :cell-count           [59 59]
   :feature-size         10
   :categories           {:sunny     {:weight     50
                                      :wind       {:min 5 :mean 10 :max 30}
                                      :temp       {:min 20 :mean 22 :max 24}
                                      :visibility {:from 15 :to 30}}
                          :fair      {:weight     50
                                      :wind       {:min 0 :mean 7 :max 20}
                                      :temp       {:min 18 :mean 21 :max 23}
                                      :visibility {:from 10 :to 30}
                                      :low-clouds {:base     {:from 3000 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :few :to :broken}}}
                          :poor      {:weight     50
                                      :wind       {:min 10 :mean 15 :max 30}
                                      :temp       {:min 15 :mean 18 :max 21}
                                      :visibility {:from 2 :to 10}
                                      :low-clouds {:base     {:from 0 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :scattered
                                                              :to   :overcast}}}
                          :inclement {:weight     50
                                      :wind       {:min 15 :mean 25 :max 60}
                                      :temp       {:min 12 :mean 14 :max 16}
                                      :visibility {:from 15 :to 30}
                                      :low-clouds {:base     {:from 0 :to 10000}
                                                   :size     {:from 0 :to 5}
                                                   :coverage {:from :scattered
                                                              :to   :overcast}}}}
   :winds-aloft          {3000  {:speed {:from 2 :to 3}
                                 :bias  0.1}
                          6000  {:speed {:from 4 :to 6}
                                 :bias  0.2}
                          9000  {:speed {:from 7 :to 9}
                                 :bias  0.3}
                          12000 {:speed {:from 8 :to 12}
                                 :bias  0.4}
                          18000 {:speed {:from 11 :to 13}
                                 :bias  0.5}
                          24000 {:speed {:from 13 :to 17}
                                 :bias  0.6}
                          30000 {:speed {:from 16 :to 18}
                                 :bias  0.7}
                          40000 {:speed {:from 18 :to 22}
                                 :bias  0.8}
                          50000 {:speed {:from 20 :to 25}
                                 :bias  0.9}}
   :turbulence           {:size 1 :power 250}
   :origin               [1000 1000]
   :evolution            3600
   :time                 {:offset  1234
                          :current {:day 1 :hour 5 :minute 0}
                          :max     nil}
   :wind-uniformity      0.7
   :crossfade            0.1
   :prevailing-wind      {:heading 325}
   :seed                 1234
   :weather-overrides    []})

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

